### PR TITLE
feat: add parameter to hide request paths in metrics

### PR DESCRIPTION
**Pull request by adamwalach**
This PR adds `hide_request_paths` parameter to remove request paths from metrics. In some scenarios, exposing paths has very big impact on metrics cardinality. 

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [ ] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [ ] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [ ] I have read the [security policy](../security/policy).
- [ ] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

**Review by aeneasr — COMMENTED** [review]
Thank you! I would actually also be fine to not make this configurable, or alternatively to use a configurable list which would allow us to specify which paths we care about. WDYT?

**Comment by adamwalach** [comment]
I don't want to introduce any regression, someone might be using the `request` label in their infrastructure. That's why the new parameter is false by default and metrics are unchanged unless someone decides otherwise. 

We already have the `collapsePaths` option enabled by default - it leaves just the root part of the URL path.
That means we could implement the list of allowed paths and in combination with `collapsePaths` we get only a handful of values which will keep cardinality of this metric on low level. 
BUT :) I think that this kind of filtering can (or should?) be implemented on WAF level, so any request that tries to access an undefined path is blocked before it reaches Oathkeeper. If we get only filtered paths, then the metric cardinality will stay low. 

In summary, I think that this PR can provide a quick solution to the metric size problem without introducing too much potentially unnecessary changes in Oathkeeper. 
But that's just my opinion :)

**Comment by aeneasr** [comment]
Thank you! 

> I don't want to introduce any regression, someone might be using the request label in their infrastructure. That's why the new parameter is false by default and metrics are unchanged unless someone decides otherwise.

From my side that would be ok, but let's make the transition smooth and remove the option at some point :)

> BUT :) I think that this kind of filtering can (or should?) be implemented on WAF level, so any request that tries to access an undefined path is blocked before it reaches Oathkeeper. If we get only filtered paths, then the metric cardinality will stay low.

The problem is that it is possible to create an infinite number of valid paths even if we filter out bad paths. That is because we have request parameters and therefore you can construct any path: `/identities/abcdef...` for example. That problem of high cardinality was already reported by users with a high volume where you have potentially millions of clients, identities, etc.

Therfore I think that if we want discoverable paths, we need to collapse them in oathkeeper or another type of filter in prometheus (or wherever).

**Comment by adamwalach** [comment]
> Thank you!
> 
> > I don't want to introduce any regression, someone might be using the request label in their infrastructure. That's why the new parameter is false by default and metrics are unchanged unless someone decides otherwise.
> 
> From my side that would be ok, but let's make the transition smooth and remove the option at some point :)
Ok, I'll create a follow-up issue about this

> > BUT :) I think that this kind of filtering can (or should?) be implemented on WAF level, so any request that tries to access an undefined path is blocked before it reaches Oathkeeper. If we get only filtered paths, then the metric cardinality will stay low.
> 
> The problem is that it is possible to create an infinite number of valid paths even if we filter out bad paths. That is because we have request parameters and therefore you can construct any path: `/identities/abcdef...` for example. That problem of high cardinality was already reported by users with a high volume where you have potentially millions of clients, identities, etc.
> 
> Therfore I think that if we want discoverable paths, we need to collapse them in oathkeeper or another type of filter in prometheus (or wherever).

We already collapse them by default: https://github.com/ory/oathkeeper/blob/master/metrics/middleware.go#L67

File: metrics/middleware.go [hunk 1]
```diff
@@ -73,6 +75,18 @@ func (m *Middleware) CollapsePaths(flag bool) *Middleware {
 	return m
 }
 
+// HidePaths if set to true, forces the value of the "request" label
+// of the prometheus request metrics to be set to an empty value.
+// eg. (when set to true):
+//   - /decisions/service/my-service -> /hidden
+//   - /decisions -> /hidden
```
Comment (aeneasr) [inline metrics/middleware.go:82]: Shouldn't this be

```suggestion
//   - /decisions/service/my-service -> /
//   - /decisions -> /
```

?

Comment (adamwalach) [reply metrics/middleware.go:82]: Right, this is a leftover from my tests. At the beginning, I used a "/hidden" value for a path to validate if I see it correctly in prometheus. I will remove the example, I think the explanation mentioning "empty value" should be sufficient.

**Review by aeneasr — APPROVED** [review]
LGTM

**Pull request merged**